cluster/addons/pod-resizer/src/nanny_lib.go

```diff
@@ -0,0 +1,81 @@
+package src
+
+import (
+	"time"
+
+	api "k8s.io/kubernetes/pkg/api/v1"
+
+	log "github.com/golang/glog"
+
+	inf "speter.net/go/exp/math/dec/inf"
+)
+
+// checkResource determines whether a specific resource needs to be over-written.
+func checkResource(threshold int64, actual, expected api.ResourceList, res api.ResourceName) bool {
+	val, ok := actual[res]
+	expVal, expOk := expected[res]
+	if ok != expOk {
+		return true
+	}
+	if !ok && !expOk {
+		return false
+	}
+	q := new(inf.Dec).QuoRound(val.Amount, expVal.Amount, 2, inf.RoundDown)
+	lower := inf.NewDec(100-threshold, 2)
+	upper := inf.NewDec(100+threshold, 2)
+	if q.Cmp(lower) == -1 || q.Cmp(upper) == 1 {
+		return true
+	}
+	return false
+}
+
+// shouldOverwriteResources determines if we should over-write the container's resource limits. We'll over-write the resource limits if the limited resources are different, or if any limit is violated by a threshold %.
+func shouldOverwriteResources(threshold int64, limits, reqs, expLimits, expReqs api.ResourceList) bool {
+	return checkResource(threshold, limits, expLimits, api.ResourceCPU) ||
+		checkResource(threshold, limits, expLimits, api.ResourceMemory) ||
+		checkResource(threshold, limits, expLimits, api.ResourceStorage) ||
+		checkResource(threshold, reqs, expReqs, api.ResourceCPU) ||
+		checkResource(threshold, reqs, expReqs, api.ResourceMemory) ||
+		checkResource(threshold, reqs, expReqs, api.ResourceStorage)
+}
+
+func PollApiServer(k8s KubernetesClient, est ResourceEstimator, contName string, pollPeriod time.Duration, threshold uint64) {
+	for i := 0; true; i++ {
+		if i != 0 {
+			// Sleep for the poll period.
+			time.Sleep(pollPeriod)
+		}
+
+		// Query the apiserver for the number of nodes.
+		num, err := k8s.CountNodes()
+		if err != nil {
+			log.Error(err)
+			continue
```

Comment:
I don't think the nanny should kill itself for attention: we should just add a log scan to the addon monitoring.

I plan to run nanny's in the same pod as the dependent container. If the nanny is in some kind of tight-kill loop, then it will drag down an important infrastructure container. If there's an issue, then we could see several pieces of critical infrastructure simultaneously fail. The operator may try to fix the churning addons, instead of root causing the real issue, the issue will be exacerbated, and we could simultaneously knock out monitoring and infrastructure (e.g., heapster and dns), which will make debugging harder.